presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +80,261 @@
+     *       calls cache.get() which returns existing queryExecutor under the
+     *          cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount != 0
+     *       releases write lock
+     *
+     * 4) Simultaneous close() and get(), close() obtains write lock first
+     *    Caller of close() obtains write lock and:
+     *       decrements the refcount
+     *       checks to see if the refcount is zero
+     *       if so, invalidates cache, closing the HTTP client as in 2 above.
+     *       releases write lock
+     *    Caller of get() obtains read lock and:
+     *       calls cache.get() which returns a new queryExecutor under the
+     *          cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *
+     * The last case of interest is the process that is using PrestoDriver
+     * calling prestoDriver.close(). This invalidates all of the cached
+     * QueryExecutors regardless of their refcount, and in doing so closes
+     * their HTTP clients. If anybody is still trying to use a
+     * PrestoConnection created by the driver, they're SOL.
+     */
     public PrestoDriver()
     {
-        this.queryExecutor = QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION);
+        this.jettyIoPool = new JettyIoPool("presto-jdbc", new JettyIoPoolConfig());
+        this.cacheLock = new ReentrantReadWriteLock();
+        this.queryExecutorCleanupService = new ScheduledThreadPoolExecutor(1);
+
+        this.queryExecutorCache = CacheBuilder.newBuilder()
+                .removalListener(new RemovalListener<Map<Object, Object>, QueryExecutorHolder>() {
+                        @Override
+                        public void onRemoval(RemovalNotification<Map<Object, Object>, QueryExecutorHolder> notification)
+                        {
+                            notification.getValue().closeQueryExecutor();
+                        }
+                    })
+                .build(new CacheLoader<Map<Object, Object>, QueryExecutorHolder>() {
+                        @Override
+                        public QueryExecutorHolder load(Map<Object, Object> clientProperties)
+                        {
+                            return new QueryExecutorHolder(
+                                    QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION, jettyIoPool),
+                                    clientProperties);
+                        }
+                    });
     }
 
     @Override
     public void close()
     {
-        queryExecutor.close();
+        if (closed.compareAndSet(false, true)) {
+            queryExecutorCleanupService.shutdown();
+            queryExecutorCache.invalidateAll();
+            jettyIoPool.close();
+        }
+    }
+
+    private class QueryExecutorHolder
```

Comment:
They are used all over the presto codebase. See Rewriters in the optimizers, for example.